src/Seabug/Contract/Util.purs

```diff
@@ -1,20 +1,209 @@
+  let
+    valHash = marketplaceValidator'.validatorHash
+    lookups = txData.lookups <> mconcat
+      [ ScriptLookups.typedValidatorLookups $ wrap marketplaceValidator'
+      , ScriptLookups.validator marketplaceValidator'.validator
+      ]
+    newNftValue =
+      Value.singleton (fst txData.newAsset) (snd txData.newAsset) one
+    constraints = txData.constraints
+      <> mustSpendScriptOutput txData.inputUtxo unitRedeemer
+      <>
+        case retBehaviour of
+          ToMarketPlace ->
+            mustPayToScript
+              valHash
+              ( Datum $ toData $
+                  MarketplaceDatum { getMarketplaceDatum: txData.newAsset }
+              )
+              newNftValue
+          ToCaller -> mempty -- Balancing will return the token to the caller
+
+  txDatumsRedeemerTxIns <- liftedE $ mkUnbalancedTx lookups constraints
+  txWithMetadata <-
+    setSeabugMetadata (modify (_ { nftId = txData.newNft }) nftData)
+      (fst txData.newAsset)
+      txDatumsRedeemerTxIns
+
+  signedTx <- liftedE
+    ( lmap
+        ( \e ->
+            name
+              <>
+                ": Cannot balance, reindex redeemers, attach datums/redeemers\
+                \ and sign: "
+              <> show e
+        )
+        <$> balanceAndSignTxE txWithMetadata
+    )
+  transactionHash <- submit signedTx
+  log $ name <> ": Transaction successfully submitted with hash: "
+    <> show transactionHash
+  log $ name <> ": Buy successful: " <> show txData.newAsset
+
+-- | Make tx data to change an nft's `NftId` by "reminting"
+-- | it. Provide a utxo map for the nft's utxo to be found in; if a
+-- | map is not provided, the current user's address will be searched.
+mkChangeNftIdTxData
+  :: forall (r :: Row Type)
+   . String
+  -> (NftId -> MintAct) -- To make the minting policy redeemer
+  -> (NftId -> NftId) -- To update the nft's `NftId`
+  -> NftData
+  -> Maybe UtxoM
+  -> Contract r SeabugTxData
+mkChangeNftIdTxData name act mapNft (NftData nftData) mScriptUtxos = do
+  let nftCollection = unwrap nftData.nftCollection
+  policy <- liftedE $ mkMintingPolicy $ wrap nftCollection
+
+  curr <- liftedM (name <> ": Cannot get CurrencySymbol")
+    $ liftAff
+    $ Value.scriptCurrencySymbol policy
+
+  let newNft = mapNft nftData.nftId
+  oldName <- liftedM (name <> ": Cannot hash old token")
+    $ mkTokenName nftData.nftId
+  newName <- liftedM (name <> ": Cannot hash new token")
+    $ mkTokenName newNft
+  let
+    oldNftValue = Value.singleton curr oldName $ negate one
+    newNftValue = Value.singleton curr newName one
+    mintRedeemer = Redeemer $ toData $ act nftData.nftId
+
+    containsNft :: forall (a :: Type). (a /\ TransactionOutput) -> Boolean
+    containsNft (_ /\ TransactionOutput out) =
+      Value.valueOf out.amount curr oldName == one
+
+  userAddr <- liftedM (name <> ": Cannot get user addr") getWalletAddress
+  userUtxos <-
+    liftedM (name <> ": Cannot get user Utxos") $ utxosAt userAddr
```

Comment:
I'm pretty sure these two queries are unnecessary when `mScriptUtxos` is not `Nothing`. The balancer will include the utxos at `getWalletAddress` anyways (and make the query anyways). Could refactor to make this query only when `mScriptUtxos` is `Nothing`, but might not be a priority to implement and test that if we're trying to get this merged quickly

Reply:
Oh good point, I'll have a look

Reply:
An untested suggestion for a starting point @samuelWilliams99 :

```diff
modified   src/Seabug/Contract/Util.purs
@@ -51,7 +51,7 @@ import Data.Array (find) as Array
 import Data.Bifunctor (lmap)
 import Data.BigInt (BigInt)
 import Data.BigInt as BigInt
-import Data.Map (insert, toUnfoldable)
+import Data.Map (singleton, toUnfoldable) as Map
 import Plutus.Types.Transaction (UtxoM)
 import Seabug.MarketPlace (marketplaceValidator)
 import Seabug.Metadata.Share (mkShare)
@@ -177,20 +177,21 @@ mkChangeNftIdTxData name act mapNft (NftData nftData) mScriptUtxos = do
     containsNft (_ /\ TransactionOutput out) =
       Value.valueOf out.amount curr oldName == one
 
-  userAddr <- liftedM (name <> ": Cannot get user addr") getWalletAddress
-  userUtxos <-
-    liftedM (name <> ": Cannot get user Utxos") $ utxosAt userAddr
-  utxo /\ utxoIndex <-
-    liftContractM (name <> ": NFT not found")
-      $ Array.find containsNft
-      $ toUnfoldable
-      $ unwrap
-      $ fromMaybe userUtxos mScriptUtxos
+    findNftUtxo = liftContractM (name <> ": NFT not found")
+      <<< Array.find containsNft
+      <<< Map.toUnfoldable
+      <<< unwrap
+
+  utxo /\ utxoIndex <- findNftUtxo =<< case mScriptUtxos of
+    Nothing -> do
+      userAddr <- liftedM (name <> ": Cannot get user addr") getWalletAddress
+      liftedM (name <> ": Cannot get user Utxos") $ utxosAt userAddr
+    Just scriptUtxos -> pure scriptUtxos
+
   let
-    utxosForTx = insert utxo utxoIndex $ unwrap userUtxos
     lookups = mconcat
       [ ScriptLookups.mintingPolicy policy
-      , ScriptLookups.unspentOutputs utxosForTx
+      , ScriptLookups.unspentOutputs $ Map.singleton utxo utxoIndex
       ]
 
     constraints = mustMintValueWithRedeemer mintRedeemer
```

Reply:
@samuelWilliams99 I could just pick this up if your testing environment isn't set up yet, might be quicker

Reply:
Oh yeah actually, that'd be great thanks!

Reply:
Implemented and tested: https://testnet.cardanoscan.io/transaction/4e306f28cde044f8c091cb808218c5b6d97d6f3c6b655d511dc4231ce5736f12

Also checked that the extra queries aren't being made